[Tooling] Fix and automate upload of Play Store metadata during release builds

AliSoftware opened the pull request:
### Description

We recently discovered that the app description and related metadata were not up-to-date in the Play Store Console, and were not matching the latest copies we had in GlotPress and in `fastlane/metadata/android/**/*.txt`

The issue were twofold:

 - There was **no** automation in place to upload the `full_description.txt` etc. files to the Play Store Console as part of the release process.
    - Apparently, up until now, this was assumed that one would run the proper `fastlane supply` command manually when new metadata were expected to be uploaded, which can easily be missed (especially if the descriptions were updated by a developer without the Release Manager knowing that new Play Store descriptions had been updated in the repo and translated)
 - Also, we were missing the `fastlane/metadata/android/en-US` folder
    - which means metadata for English would not have been included in the automatic upload of the metadata by the `fastlane supply` (aka `fastlane upload_to_play_store`) action anyway.

### Solution 

- Added missing `en-US` folder and its `changelogs/` subfolder in `fastlane/metadata/android`, with a `.gitkeep` file to make sure it's kept around even if empty
- Fix `download_metadata_strings` for the `en-US` case, to make sure `.txt` files used as the source of truth for translation are copied into `fastlane/metadata/android/en-US` properly
- Finally, make our `build_and_upload_release` lane also upload metadata and release notes to PlayStore automatically, by turning `skip_upload_metadata` and `skip_upload_changelogs` to `false` during the upload of a new build to the production track.

### Testing instructions

 - [x] Run `bundle exec fastlane download_metadata_strings` and confirm that at the end of it, all the `*.txt` files in `fastlane/metadata/android/**.*.txt` match our expectation, and especially that it is now including the English source copy in `en-US/` subfolder
 - [ ] During the next release of WCAndroid on Friday, check that the Release Build CI step—which ends up calling `build_and_upload_google_play` which in turn is the one that calls `build_and_upload_release`—will properly update the metadata and release notes for you

> **Note**: I've only tested the first step, leaving the final test to you @ParaskP7  on Friday.

Actually you could consider testing that the update of the metadata work by using `upload_to_play_store` even before Friday (to avoid risk of last-minute surprises about translations too long or whatnot) by calling `upload_to_play_store` with just `skip_upload_metadata: false` and all the other `skip_*` to true, allowing you to just run that action to upload metadata even without having to wait for Friday or having to upload a new AAB or creating any new release.

You should be able to do that by adding a temporary dummy lane to your `Fastfile` like this:

```ruby
lane :upload_metadata_to_play_store do
  upload_to_play_store(
    package_name: APP_PACKAGE_NAME,
    skip_upload_aab: true,
    skip_upload_apk: true,
    skip_upload_metadata: false,
    skip_upload_changelogs: true,
    skip_upload_images: true,
    skip_upload_screenshots: true,
    json_key: UPLOAD_TO_PLAY_STORE_JSON_KEY,
    # Retain the build for v10.9.2 which is the latest one supporting Android 5/6
    version_codes_to_retain: [356],
  )
end
```

Then `bundle exec fastlane upload_metadata_to_play_store` and go check on Google Play Console to validate it worked. Or if the command failed on some locales saying the translation copies exceeded the max character limit, that will give you enough time to reach out to +i18n to ask for translators to provide shorter translations before Friday.

(@ParaskP7 I haven't done that test step on my end either, I'll let you run it and handle the follow-up with +18n if one is needed)

### Next Step

With this change, the release notes XML should already be automatically pre-populated by the `upload_to_play_store(…, skip_upload_changelog: false, …)` during CI release build.
Which means that we should later be able to get rid of:
 - The calls to `android_create_xml_release_notes` + `add_us_release_notes`, which are used to generate the `release_notes.xml` file
 - The `release_notes.xml` file itself, which we could delete as it won't be necessary nor generated anymore
 - The step in the WCAndroid release scenario that tells you to manuallly copy/paste the `release_notes.xml` into Play Store Console

@ParaskP7 I've left those two changes out of this PR just to make sure you have the opportunity to first do the next release of WCAndroid to confirm that the changes in this PR work. If you confirm they do work as expected, then feel free to create a subsequent PR next week to get rid of those (similar to what was done in https://github.com/wordpress-mobile/WordPress-Android/pull/16281)

AliSoftware committed:
Add missing en-US folder for metadata

AliSoftware committed:
Fix `download_metadata_strings` for the `en-US` case

AliSoftware committed:
Make `build_and_upload_release` also upload metadata and release notes to PlayStore automatically

AliSoftware committed:
Update metadata translations for 12.5

ParaskP7 reviewed (commented):
👋 @AliSoftware !

Thank YOU so much for opening this PR and for the very useful and super-detailed description, as always, you rock! 🥇 💯 🚀 

I reviewed and tested this PR as per your testing instructions, excluding testing the `upload_metadata_to_play_store` dummy lane (more on that later). Just as you did, testing `download_metadata_strings` worked as you described. 👍 

-----

> ... with a .gitkeep file to make sure it's kept around even if empty

Question (❓):  Do we really need these 2 x `.gitkeep` files? PS: I don't see them on WPAndroid. 🤔 

-----

> Actually you could consider testing that the update of the metadata work by using upload_to_play_store even before Friday (to avoid risk of last-minute surprises about translations too long or whatnot) by calling upload_to_play_store with just skip_upload_metadata: false and all the other skip_* to true, allowing you to just run that action to upload metadata even without having to wait for Friday or having to upload a new AAB or creating any new release.

@hichamboushaba I am considering testing that, but, I am not sure if you are about to do that as well (manually), wdyt, should we test that and see if the `metadata` gets updated automatically for us instead of doing it manually? 🤔 

@AliSoftware just to be 100% sure, if I run this `upload_metadata_to_play_store` dummy lane, this will change the `title`, `short/full description` on WCAndroid main store listing, thus update our production PlayStore values, right? Doing that wouldn't allow us to review the change within Google Play Console and then submit/abort that in case something goes wrong, instead it will do the update directly, right? 🤔

-----

Suggestion (💡): @AliSoftware since I am seeing the `metadata` in this PR as half-baked (since the translations for those wasn't 100% ready yesterday), but also, noticing that you used `388` as the `build_number` instead of ` 389`, which is what usually happens as we first bump the version name/code before running this lane, I thus suggest that I cherry-pick those commit of yours tomorrow during the `12.5` release finalization instead (or even delay that for next release, next Friday), and then get the `Update metadata translations for 12.5` commit be generated as usual via `finalize_release`, wdyt?

-----

> Next Step

Thanks for sharing the next steps too, I agree, those should be done after we verify that the above process works as expected and that there isn't anything else we are missing there. 💯

hichamboushaba commented:
> @hichamboushaba I am considering testing that, but, I am not sure if you are about to do that as well (manually), wdyt, should we test that and see if the metadata gets updated automatically for us instead of doing it manually? 🤔

@ParaskP7 I agree, let's test it, if it works then it's better than having to manually update them, and it'll avoid the issue next time.

ParaskP7 commented:
> @ParaskP7 I agree, let's test it, if it works then it's better than having to manually update them, and it'll avoid the issue next time.

Great, I'll test it then now and report back here @hichamboushaba ! 👍

But, just to make sure of it, to my understanding, me running this `upload_metadata_to_play_store` lane will, most probably, if everything goes as expected, update all of WPAndroid PlayStore listings (`title` + `short/full description`), getting them from this PR (aka `trunk`), are you sure you want me doing that for you right now?

PS: Also, from what I checked you haven't manually updated any of those PlayStore listings yourself, please correct me if I am wrong.

AliSoftware commented:
> Question (❓): Do we really need these 2 x .gitkeep files? PS: I don't see them on WPAndroid. 🤔

Probably not indeed, now that those folders are populated with some files (that might get their content changed each sprint), I guess there's no real risk anymore to have those folders end up empty and risk not being seen to git anymore. 🤷 

> just to be 100% sure, if I run this upload_metadata_to_play_store dummy lane, this will change the title, short/full description on WCAndroid main store listing, thus update our production PlayStore values, right? Doing that wouldn't allow us to review the change within Google Play Console and then submit/abort that in case something goes wrong, instead it will do the update directly, right?

Correct. Though those changes to title/short/long desc still go thru approval (i.e. you'll see them under "Changes in review" in the "Publishing Overview" section of the Google Play Console, waiting to be reviewed and approved by Google, as opposed to them going live _immediately_. But that dummy lane will indeed submit them for review directly without allowing you to check the result first before submitting.

But! Looking at the documentation for `upload_to_play_store` (which you can see by running `bundle exec fastlane action upload_to_play_store`) revealed that there's a `validate_only: true` parameter that you can probably use here in your dummy lane, to validate without publishing.
So adding that `validate_only: true` parameter to the `upload_to_play_store` call in your dummy lane should be a good way to debug this before setting it back to `false` and going all in 🙂

ParaskP7 commented:
> Probably not indeed, now that those folders are populated with some files (that might get their content changed each sprint), I guess there's no real risk anymore to have those folders end up empty and risk not being seen to git anymore. 🤷

👍 

> Correct. Though those changes to title/short/long desc still go thru approval (i.e. you'll see them under "Changes in review" in the "Publishing Overview" section of the Google Play Console, waiting to be reviewed and approved by Google, as opposed to them going live immediately. But that dummy lane will indeed submit them for review directly without allowing you to check the result first before submitting.

Thanks for the clarification on that @AliSoftware ! 🙏 

> But! Looking at the documentation for upload_to_play_store (which you can see by running bundle exec fastlane action upload_to_play_store) revealed that there's a validate_only: true parameter that you can probably use here in your dummy lane, to validate without publishing.
So adding that validate_only: true parameter to the upload_to_play_store call in your dummy lane should be a good way to debug this before setting it back to false and going all in 🙂

Oh, nice find about the `validate_only: true` parameter, I'll indeed use that first with this dummy lane before going all-in! 🥇 👌

AliSoftware commented:
> Suggestion (💡): @AliSoftware since I am seeing the metadata in this PR as half-baked (since the translations for those wasn't 100% ready yesterday), but also, noticing that you used 388 as the build_number instead of  389, which is what usually happens as we first bump the version name/code before running this lane, I thus suggest that I cherry-pick those commit of yours tomorrow during the 12.5 release finalization instead (or even delay that for next release, next Friday), and then get the Update metadata translations for 12.5 commit be generated as usual via finalize_release, wdyt?

Up to you.

 - Either you merge this PR as is, well probably after rebasing it to `release/12.5`, `388.txt` and all
    - Then when you'll run `finalize_release` on Friday, the `changelogs/388.txt` files added by that PR will be deleted again, and replaced with the new build number used for the final release (`389.txt`, probably), also including all the latest translations from GlotPress that are not yet 100% today
 - Or, while you rebase my PR onto `release/12.5`, you only pick the changes in the `Fastfile` + the `.gitkeep` files (basically discarding commit 2c09130 from this PR during rebase)
    - Thus merging only the changes in the tooling scripts and discarding the updates in the `changelogs/*.txt` and `release_notes.xml` from that PR, keeping those unchanged (and keeping the `en-US` and `en-US/changelogs` folders existant… but empty for now).
   - Then those changelogs and metadata translations will be only updated on Friday when you run `finalize_release`
 - As a middle ground, you could also consider only discarding the `changelog/388.txt` files from this PR… but keeping the translations for the `{title,short_description,long_description}.txt`
    - This will allow you to use the dummy lane to test/upload those metadata translations before Friday).
    - Since the main inconsistent thing is the build number used for those `changelogs/*.txt` of that PR not being the final one, maybe that one is the best of both worlds? 🤷 

But to be honest, commiting those `388.txt` from that PR wouldn't really be a big deal, because (1) the dummy lane will not use them as long as you use `skip_upload_changelogs: true` (while title/description metadata is controlled separately, by `skip_upload_metadata: false`) and (2) those will be deleted and replaced by the changelogs with new build number when you'll run `finalize_release` on Friday anyway. And even if you used `skip_upload_changelogs: false`, that would only have updated the changelog of the release attached to the build with `388` version Code (which is probably a beta build in the Open Testing track, not in Production track… so… wouldn't even be a big deal)

hichamboushaba commented:
> But, just to make sure of it, to my understanding, me running this upload_metadata_to_play_store lane will, most probably, if everything goes as expected, update all of WPAndroid PlayStore listings (title + short/full description), getting them from this PR (aka trunk), are you sure you want me doing that for you right now?

Thanks @ParaskP7, I just double-checked, and the `title` and `short_description` in the console match the values in the repo, so the update won't impact them.
And I confirm that the update for the `full_description` is indeed what we want to do.

> PS: Also, from what I checked you haven't manually updated any of those PlayStore listings yourself, please correct me if I am wrong.

Yes, I never updated them, I updated only the screenshots manually few months ago.

ParaskP7 commented:
FYI @hichamboushaba @AliSoftware that I just now did try and run `bundle exec fastlane upload_metadata_to_play_store validate_only:true` after adding this dummy lane in. As I (kind of) expected, I am getting the following `insufficient permissions` error:

```
[15:22:12]: Google Api Error: Invalid request - The current user has
insufficient permissions to perform the requested operation.
```

<details>
  <summary>Full Log</summary>

```
[15:21:58]: ----------------------------------
[15:21:58]: --- Step: upload_to_play_store ---
[15:21:58]: ----------------------------------

+------------------------------------+------------------------------------------------------------------------------------+
|                                               Summary for supply 2.210.1                                                |
+------------------------------------+------------------------------------------------------------------------------------+
| package_name                       | com.woocommerce.android                                                            |
| skip_upload_aab                    | true                                                                               |
| skip_upload_apk                    | true                                                                               |
| skip_upload_metadata               | false                                                                              |
| skip_upload_changelogs             | true                                                                               |
| skip_upload_images                 | true                                                                               |
| skip_upload_screenshots            | true                                                                               |
| json_key                           | /Users/petros.paraskevopoulos/.configure/woocommerce-android/secrets/google-uploa  |
|                                    | d-credentials.json                                                                 |
| version_codes_to_retain            | [356]                                                                              |
| release_status                     | completed                                                                          |
| track                              | production                                                                         |
| metadata_path                      | ./fastlane/metadata/android                                                        |
| track_promote_release_status       | completed                                                                          |
| validate_only                      | false                                                                              |
| check_superseded_tracks            | false                                                                              |
| timeout                            | 300                                                                                |
| deactivate_on_promote              | true                                                                               |
| changes_not_sent_for_review        | false                                                                              |
| rescue_changes_not_sent_for_review | true                                                                               |
| ack_bundle_installation_warning    | false                                                                              |
+------------------------------------+------------------------------------------------------------------------------------+

[15:21:59]: Updating track 'production'...
[15:22:00]: Preparing to upload for language 'ar'...
[15:22:01]: Preparing to upload for language 'de-DE'...
[15:22:02]: Preparing to upload for language 'en-US'...
[15:22:02]: Preparing to upload for language 'es-ES'...
[15:22:03]: Preparing to upload for language 'fr-FR'...
[15:22:03]: Preparing to upload for language 'id'...
[15:22:04]: Preparing to upload for language 'it-IT'...
[15:22:04]: Preparing to upload for language 'iw-IL'...
[15:22:05]: Preparing to upload for language 'ja-JP'...
[15:22:06]: Preparing to upload for language 'ko-KR'...
[15:22:06]: Preparing to upload for language 'nl-NL'...
[15:22:07]: Preparing to upload for language 'pt-BR'...
[15:22:07]: Preparing to upload for language 'ru-RU'...
[15:22:08]: Preparing to upload for language 'sv-SE'...
[15:22:09]: Preparing to upload for language 'tr-TR'...
[15:22:09]: Preparing to upload for language 'zh-CN'...
[15:22:10]: Preparing to upload for language 'zh-TW'...
[15:22:11]: Uploading all changes to Google Play...
+------------------+---------------------------------------+
|                       Lane Context                       |
+------------------+---------------------------------------+
| DEFAULT_PLATFORM | android                               |
| PLATFORM_NAME    | android                               |
| LANE_NAME        | android upload_metadata_to_play_store |
+------------------+---------------------------------------+
[15:22:12]: Google Api Error: Invalid request - The current user has insufficient permissions to perform the requested operation.

+------+---------------------------+-------------+
|                fastlane summary                |
+------+---------------------------+-------------+
| Step | Action                    | Time (in s) |
+------+---------------------------+-------------+
| 1    | default_platform          | 0           |
| 2    | is_ci                     | 0           |
| 3    | check_for_toolkit_updates | 6           |
| 💥   | upload_to_play_store      | 13          |
+------+---------------------------+-------------+

[15:22:12]: fastlane finished with errors

[!] Google Api Error: Invalid request - The current user has insufficient permissions to perform the requested operation.
```

</details>

Maybe one of you could try and run this on you local machines instead?

Btw @AliSoftware is that because I anyway don't have permission to do that, even manually via the Google Play Console UI (thus I might need to request elevated access to systems for WCAndroid), or is it because I haven't (somehow) configured this `google-upload-credentials.json` file which is used with `UPLOAD_TO_PLAY_STORE_JSON_KEY`? 🤔

```
You currently don’t have the necessary rights to edit this page.
Please ask the account owner to grant you access.
```

AliSoftware commented:
The `UPLOAD_TO_PLAY_STORE_JSON_KEY` points to a file in `.configure`, so running `bundle exec fastlane run configure_apply` should get you all you need?

AliSoftware commented:
Mmmh ok, I got the same Google Api permissions error on my end 😅  We'll have to dig deeper to check if it's a missing role in the credentials/key we have in mobile secrets for WCAndroid, for which we'd need to ask Systems to elevate permissions for it, or if it's just something local (like credentials not configured on our Mac but would work on CI).

I've been away from the Woo codebase for too long to remember tbh, so not sure, could just be us missing a configuration step on our Macs to use the proper credentials

AliSoftware commented:
@ParaskP7 I have to get moving to go back to my other project and Tumblr Release duties for the day.
Can I let you check maybe with @oguzkocer about this Google Api permission thing (checking if it's only a config missing on our Macs, or if it's because the key we have in mobile secrets have insufficient permissions / roles)?

ParaskP7 commented:
> Mmmh ok, I got the same Google Api permissions error on my end 😅

😅 

> We'll have to dig deeper to check if it's a missing role in the credentials/key we have in mobile secrets for WCAndroid, for which we'd need to ask Systems to elevate permissions for it, or if it's just something local (like credentials not configured on our Mac but would work on CI).

I just requested store listing access from systems (copy-pasted @hichamboushaba request, thanks) and I'll try that again after I get that sorted, maybe that will do the trick... 🤷 

> I've been away from the Woo codebase for too long to remember tbh, so not sure, could just be us missing a configuration step on our Macs to use the proper credentials

🤷 🤔 🤷 To my understanding CI was doing the upload and maybe that configuration and credentials are configured on CI only... 🤷 🤔 🤷

Just to clarify @AliSoftware , you could do exactly that on WPAndroid, right?

ParaskP7 commented:
> @ParaskP7 I have to get moving to go back to my other project and Tumblr Release duties for the day.

Of course, PLEASE and THANK you SO MUCH for ALL the HELP and SUPPORT you provided ALREADY. 🙇 🙇 🙇

> Can I let you check maybe with @oguzkocer about this Google Api permission thing (checking if it's only a config missing on our Macs, or if it's because the key we have in mobile secrets have insufficient permissions / roles)?

👍 💯 🙏

ParaskP7 commented:
> Thanks @ParaskP7, I just double-checked, and the title and short_description in the console match the values in the repo, so the update won't impact them.
And I confirm that the update for the full_description is indeed what we want to do.

Thanks for the clarification on that @hichamboushaba ! 🙇 

> Yes, I never updated them, I updated only the screenshots manually few months ago.

Got it, yes, I was just trying to make sure you didn't do the update yesterday or about to do it today, thus me potentially updating you update somehow while trying the dummy lane.

AliSoftware commented:
- Same dummy lane works on WPAndroid when run locally (after replacing the `APK_PACKAGE_NAME` constant with `org.wordpress.android`) without any credentials error.
 - What's strange though is that the key file (pointing by `UPLOAD_TO_PLAY_STORE_JSON_KEY` in each case, which is the  `$HOME/.configure/{repo}/secrets/google-upload-credentials.json` file that comes from the corresponding secrets from `.mobile-secrets`) has the same content for both repos' secrets:

```
$ diff ~/.configure/{wordpress,woocommerce}-android/secrets/google-upload-credentials.json && echo "Same"
Same
```
So that means the difference probably lies in the scopes/permissions that this token (which is NOT your personal token but a key used by CI and shared by projects) has: probably, that JSON key have the sufficient/needed scopes for the WPAndroid app on GPC but not for the WCAndroid app? (i.e. I'm assuming the scopes of this token are defined separately per app which would explain this?).

> I just requested store listing access from systems (copy-pasted @hichamboushaba request, thanks) and I'll try that again after I get that sorted

If that theory above (about separate scopes for WP/WC given by the token) is true, Systems will have to update that the scopes/permissions of specific token used by our CI and automation, not your personal permissions (which are only relevant when you do things manually on the Console's web UI while connected with your own account)

ParaskP7 commented:
> Up to you.

From your [above](https://github.com/woocommerce/woocommerce-android/pull/8439#issuecomment-1451862267) answer @AliSoftware , I think I will still go with cherry-picking your commits instead, or even redoing them in order to only keep those files that are actually needed, and doing all that on the current active `release/12.5` branch tomorrow before me starting with the release finalization steps for `12.5`.

However, I would only do that after I verify that this dummy `upload_metadata_to_play_store` lane works as expected, thus after me testing that this solution. I want to avoid any hiccups tomorrow during finalization. In the worst case, I'll ignore these changes on this release cycle, do the finalization as usual, and start fresh with the changes on Monday, testing and all.

PS: @hichamboushaba that in the worse case scenario I described above, me not being able to test that this dummy `upload_metadata_to_play_store` lane works as expected, if you are in a hurry to update those store listing, we could do that manually tomorrow.

ParaskP7 commented:
> Same dummy lane works on WPAndroid when run locally (after replacing the APK_PACKAGE_NAME constant with org.wordpress.android) without any credentials error.

Good to know! 👍 

> What's strange though is that the key file (pointing by UPLOAD_TO_PLAY_STORE_JSON_KEY in each case, which is the $HOME/.configure/{repo}/secrets/google-upload-credentials.json file that comes from the corresponding secrets from .mobile-secrets) has the same content for both repos' secrets:

👍

> So that means the difference probably lies in the scopes/permissions that this token (which is NOT your personal token but a key used by CI and shared by projects) has

👍

> probably, that JSON key have the sufficient/needed scopes for the WPAndroid app on GPC but not for the WCAndroid app? (i.e. I'm assuming the scopes of this token are defined separately per app which would explain this?).

👍 🤔 

> If that theory above (about separate scopes for WP/WC given by the token) is true, Systems will have to update that the scopes/permissions of specific token used by our CI and automation, not your personal permissions (which are only relevant when you do things manually on the Console's web UI while connected with your own account)

Got it, this makes sense @AliSoftware ! 👍 

> Systems will have to update that the scopes/permissions of specific token used by our CI and automation

- In case your theory is correct, any idea on how to go about this with systems, create a request of some kind, do we have an example request like that? 🤔
- If that theory doesn't stand, what could be the next step for us to move forward from here? 🤔

PS: I know you need to get moving with your work for today, so please feel free to answer to this when you can.

ParaskP7 commented:
Update: @hichamboushaba @AliSoftware

> ...not your personal permissions (which are only relevant when you do things manually on the Console's web UI while connected with your own account)

I got my permissions elevated by systems and as Olivier correctly logically pointed out I still have `insufficient permissions` when running this dummy lane of ours locally. 👍 

As such, at this point, I am blocked on testing this locally. 🤷 

> During the next release of WCAndroid on Friday, check that the Release Build CI step—which ends up calling build_and_upload_google_play which in turn is the one that calls build_and_upload_release—will properly update the metadata and release notes for you

As such, as per this testing instruction above, an FYI to you that I'll try the update tomorrow during the finalization of `12.5` and make sure that when the release gets build on CI it will properly update the metadata and the release notes for us.

I think this is the best we can do right now. 🤷 

And if, for whatever reason, that won't work tomorrow as well, even with CI's help, then I'll just revert my cherry-picked commits from Olivier and progress without this solution in place. Then, we will update those store listing manually this time and until we figure out how to automated the WCAndroid release to do so for us for next time.

> In case your theory is correct, any idea on how to go about this with systems, create a request of some kind, do we have an example request like that? 🤔
> If that theory doesn't stand, what could be the next step for us to move forward from here? 🤔

No matter, as a subsequent step to the all the above, it would be great to be able to run these kind of play store related lanes locally too, so I suggest we try and figure out the root cause of why we can't run those locally on WCAndroid, just like we are able to do on WPAndroid.

AliSoftware commented:
You can always temporarily commit the dummy lane on top of this PR's branch + add a call to it in `.buildkite/pipeline.yml` and see how it goes and if CI would pass.

Though I doubt it because I don't see any specific secrets related to upload keys / credentials in `~/.mobile-secrets/CI/secrets/{wc,wp}android/env` that would allow CI to have different credentials than the one we have locally from running `configure_apply`.

Besides, it would still be a good thing to be able to run that locally (eg in case during a sprint there's an issue in CI and you have to work around it etc)

I think just create a systemsrequest to explain that you need the key to have permissions to update metadata on WC like it has on WP, and see how the conversation goes with Systems go on that P2 from here. Can't hurt to ask.

oguzkocer commented:
Re: API permission error

@ParaskP7 I looked at your systems request in which you mention your personal email and request that to be granted permission for editing the metadata for WCAndroid. I don't think that's correct. As @AliSoftware previously pointed out, the credentials used to submit the app & metadata to Play Store is a shared credential that lives in `.mobile-secrets`. Unless I am missing something, the proper request would need to ask that those shared credentials to be granted access for editing metadata. Also, as again mentioned by @AliSoftware, this shared credential is also used in WPAndroid and has the proper permissions, so sharing that information with systems should be helpful.

Just to be extra clear, the shared credential in question is the [`google-upload-credentials.json` file](https://github.com/woocommerce/woocommerce-android/blob/trunk/fastlane/Fastfile#L68) that's copied by the `configure_apply` [here](https://github.com/woocommerce/woocommerce-android/blob/trunk/.configure#L36-L40). Note that there is also a [`google-services.json` file](https://github.com/woocommerce/woocommerce-android/blob/trunk/.configure#L11-L15) but that's NOT the one that's relevant to uploading to Play Store and it's instead used for [Google Services](https://developers.google.com/android/guides/google-services-plugin) within the [app](https://github.com/woocommerce/woocommerce-android/blob/trunk/WooCommerce/build.gradle#L13).

I hope that helps 🤞

ParaskP7 commented:
👋 @AliSoftware !

> You can always temporarily commit the dummy lane on top of this PR's branch + add a call to it in .buildkite/pipeline.yml and see how it goes and if CI would pass.

👍 

> Though I doubt it because I don't see any specific secrets related to upload keys / credentials in ~/.mobile-secrets/CI/secrets/{wc,wp}android/env that would allow CI to have different credentials than the one we have locally from running configure_apply.

👍 

> Besides, it would still be a good thing to be able to run that locally (eg in case during a sprint there's an issue in CI and you have to work around it etc)

💯 

> I think just create a systemsrequest to explain that you need the key to have permissions to update metadata on WC like it has on WP, and see how the conversation goes with Systems go on that P2 from here. Can't hurt to ask.

That's my next step! 👍 

FYI: As per the above and the conversations we had so far, it seems that me running this lane locally should be all we need to verify that this solution works, no point to update `12.5` and try that on CI at this point. As such, I'll proceed with today's WCAndroid release without this solution in place and only incorporate that when we get this shorted out with the help from systems.

PS: @hichamboushaba if you are in a hurry on updating the store listing I suggest you doing it manually this time. This is because of the fact that I am unsure when this automatic playstore upload functionality will be ready for us to use, it might get shorted quickly with a simple system request, but it might as well not as they might be other blockers to it as well, I am just not sure at this point.

ParaskP7 commented:
👋 @oguzkocer and thanks for your input!

> @ParaskP7 I looked at your systems request in which you mention your personal email and request that to be granted permission for editing the metadata for WCAndroid. I don't think that's correct.

Yes, that's true, as per the all the above discussions we had, I just thought to try that anyway, just in case we are missing something, but that was a long shot and I was aware of my chances it not being what's blocking me. 😅

> As @AliSoftware previously pointed out, the credentials used to submit the app & metadata to Play Store is a shared credential that lives in .mobile-secrets.

💯

> Unless I am missing something, the proper request would need to ask that those shared credentials to be granted access for editing metadata.

That's my next step! 👍

> Also, as again mentioned by @AliSoftware, this shared credential is also used in WPAndroid and has the proper permissions, so sharing that information with systems should be helpful.

💯

ParaskP7 commented:
> That's my next step! 👍

@AliSoftware @oguzkocer @hichamboushaba FYI: I have made the system request and now waiting on their response (`pMz3w-h2B-p2`). 🤞 

PS: Feel free to add anything in that request that I might have missed. 💯

ParaskP7 commented:
> @AliSoftware @oguzkocer @hichamboushaba FYI: I have made the system request and now waiting on their response (pMz3w-h2B-p2). 🤞

@AliSoftware @oguzkocer @hichamboushaba FYI: After system granted this permission to us, I verified that by running `bundle exec fastlane upload_metadata_to_play_store validate_only:true`, which completed successfully.

We are now able to to automatically update the store listing of the WooCommerce app in Google Play. 🎉

-----

> But! Looking at the documentation for upload_to_play_store (which you can see by running bundle exec fastlane action upload_to_play_store) revealed that there's a validate_only: true parameter that you can probably use here in your dummy lane, to validate without publishing.
So adding that validate_only: true parameter to the upload_to_play_store call in your dummy lane should be a good way to debug this before setting it back to false and going all in 🙂

@AliSoftware that me adding the `validate_only:true` and running the lane actually updated the store listing, thus this parameter might not be doing what we thought it would... 🤷 

<details>
    <summary>bundle exec fastlane upload_metadata_to_play_store validate_only:true</summary>

```
petros.paraskevopoulos@Mac-Studio woocommerce-android % bundle exec fastlane upload_metadata_to_play_store validate_only:true
[✔] 🚀 
+-----------------------------------+---------+-------------------------------------------------------+
|                                            Used plugins                                             |
+-----------------------------------+---------+-------------------------------------------------------+
| Plugin                            | Version | Action                                                |
+-----------------------------------+---------+-------------------------------------------------------+
| fastlane-plugin-wpmreleasetoolkit | 6.3.0   | configure_download, configure_setup,                  |
|                                   |         | configure_apply, configure_update,                    |
|                                   |         | configure_add_files_to_copy, configure_validate,      |
|                                   |         | ios_download_strings_files_from_glotpress,            |
|                                   |         | ios_generate_strings_file_from_code,                  |
|                                   |         | ios_build_prechecks, ios_betabuild_prechecks,         |
|                                   |         | ios_build_preflight, ios_localize_project,            |
|                                   |         | ios_merge_strings_files, ios_update_release_notes,    |
|                                   |         | add_devices_to_provisioning_profiles,                 |
|                                   |         | ios_current_branch_is_hotfix,                         |
|                                   |         | add_development_certificates_to_provisioning_profile  |
|                                   |         | s, ios_tag_build, ios_final_tag,                      |
|                                   |         | ios_finalize_prechecks, ios_bump_version_hotfix,      |
|                                   |         | ios_check_beta_deps, ios_validate_ci_build,           |
|                                   |         | ios_get_build_version, ios_get_app_version,           |
|                                   |         | ios_lint_localizations, ios_codefreeze_prechecks,     |
|                                   |         | ios_extract_keys_from_strings_files,                  |
|                                   |         | ios_update_metadata, ios_clear_intermediate_tags,     |
|                                   |         | ios_bump_version_beta,                                |
|                                   |         | ios_completecodefreeze_prechecks,                     |
|                                   |         | ios_send_app_size_metrics,                            |
|                                   |         | ios_update_metadata_source,                           |
|                                   |         | ios_bump_version_release, ios_hotfix_prechecks,       |
|                                   |         | ios_get_store_app_sizes, circleci_trigger_job,        |
|                                   |         | check_translation_progress, firebase_login,           |
|                                   |         | upload_to_s3, extract_release_notes_for_version,      |
|                                   |         | gp_downloadmetadata, create_release, setfrozentag,    |
|                                   |         | comment_on_pr, close_milestone,                       |
|                                   |         | buildkite_trigger_build, promo_screenshots,           |
|                                   |         | removebranchprotection, gp_update_metadata_source,    |
|                                   |         | get_prs_list, create_new_milestone,                   |
|                                   |         | check_for_toolkit_updates, setbranchprotection,       |
|                                   |         | android_completecodefreeze_prechecks,                 |
|                                   |         | android_download_file_by_version,                     |
|                                   |         | android_download_translations,                        |
|                                   |         | android_betabuild_prechecks,                          |
|                                   |         | android_bump_version_release,                         |
|                                   |         | android_send_app_size_metrics,                        |
|                                   |         | android_current_branch_is_hotfix,                     |
|                                   |         | android_build_prechecks,                              |
|                                   |         | android_bump_version_final_release,                   |
|                                   |         | android_firebase_test, android_build_preflight,       |
|                                   |         | android_get_release_version,                          |
|                                   |         | android_codefreeze_prechecks, android_tag_build,      |
|                                   |         | android_create_xml_release_notes, an_localize_libs,   |
|                                   |         | android_bump_version_hotfix,                          |
|                                   |         | android_launch_emulator,                              |
|                                   |         | android_update_release_notes,                         |
|                                   |         | android_shutdown_emulator, android_create_avd,        |
|                                   |         | an_validate_lib_strings, android_get_alpha_version,   |
|                                   |         | android_get_app_version, android_finalize_prechecks,  |
|                                   |         | an_update_metadata_source, android_hotfix_prechecks,  |
|                                   |         | android_bump_version_beta                             |
+-----------------------------------+---------+-------------------------------------------------------+

[12:02:07]: ------------------------------
[12:02:07]: --- Step: default_platform ---
[12:02:07]: ------------------------------
[12:02:07]: Driving the lane 'android upload_metadata_to_play_store' 🚀
[12:02:07]: -------------------
[12:02:07]: --- Step: is_ci ---
[12:02:07]: -------------------
[12:02:07]: ---------------------------------------
[12:02:07]: --- Step: check_for_toolkit_updates ---
[12:02:07]: ---------------------------------------
[12:02:07]: Currently using release toolkit version 6.3.0.
[12:02:07]: Checking for updates now...
[12:02:09]: There is a newest version 7.0.0 of the release toolkit!
[12:02:09]: The latest version available (7.0.0) introduces breaking changes compared to the 6.3.0 you are currently using.

 - To update to 7.0.0, first edit your `Pluginfile` to use '~> 7.0', run `bundle update`,
   then be sure to make all the necessary changes to your `Fastfile` (see the toolkit's CHANGELOG)
   to take those breaking changes into account.

 - If you are not ready to make the major version bump, you can still try to update to the latest compatible,
   non-breaking version by running `bundle update` now. This will not update to the latest 7.0.0, but
   might still update to a newer version compatible with '~> 6.3' if one exists; which is still valuable
   to at least get bugfixes, until you are ready to jump to the next major version later.

[12:02:09]: Do you want to run bundle update now? (y/n)
n
[12:02:12]: ----------------------------------
[12:02:12]: --- Step: upload_to_play_store ---
[12:02:12]: ----------------------------------

+------------------------------------+------------------------------------------------------------------------------------+
|                                               Summary for supply 2.210.1                                                |
+------------------------------------+------------------------------------------------------------------------------------+
| package_name                       | com.woocommerce.android                                                            |
| skip_upload_aab                    | true                                                                               |
| skip_upload_apk                    | true                                                                               |
| skip_upload_metadata               | false                                                                              |
| skip_upload_changelogs             | true                                                                               |
| skip_upload_images                 | true                                                                               |
| skip_upload_screenshots            | true                                                                               |
| json_key                           | /Users/petros.paraskevopoulos/.configure/woocommerce-android/secrets/google-uploa  |
|                                    | d-credentials.json                                                                 |
| version_codes_to_retain            | [356]                                                                              |
| release_status                     | completed                                                                          |
| track                              | production                                                                         |
| metadata_path                      | ./fastlane/metadata/android                                                        |
| track_promote_release_status       | completed                                                                          |
| validate_only                      | false                                                                              |
| check_superseded_tracks            | false                                                                              |
| timeout                            | 300                                                                                |
| deactivate_on_promote              | true                                                                               |
| changes_not_sent_for_review        | false                                                                              |
| rescue_changes_not_sent_for_review | true                                                                               |
| ack_bundle_installation_warning    | false                                                                              |
+------------------------------------+------------------------------------------------------------------------------------+

[12:02:14]: Updating track 'production'...
[12:02:15]: Preparing to upload for language 'ar'...
[12:02:16]: Preparing to upload for language 'de-DE'...
[12:02:17]: Preparing to upload for language 'en-US'...
[12:02:18]: Preparing to upload for language 'es-ES'...
[12:02:18]: Preparing to upload for language 'fr-FR'...
[12:02:19]: Preparing to upload for language 'id'...
[12:02:20]: Preparing to upload for language 'it-IT'...
[12:02:21]: Preparing to upload for language 'iw-IL'...
[12:02:22]: Preparing to upload for language 'ja-JP'...
[12:02:23]: Preparing to upload for language 'ko-KR'...
[12:02:23]: Preparing to upload for language 'nl-NL'...
[12:02:24]: Preparing to upload for language 'pt-BR'...
[12:02:25]: Preparing to upload for language 'ru-RU'...
[12:02:25]: Preparing to upload for language 'sv-SE'...
[12:02:26]: Preparing to upload for language 'tr-TR'...
[12:02:27]: Preparing to upload for language 'zh-CN'...
[12:02:28]: Preparing to upload for language 'zh-TW'...
[12:02:28]: Uploading all changes to Google Play...
[12:02:31]: Successfully finished the upload to Google Play

+------+---------------------------+-------------+
|                fastlane summary                |
+------+---------------------------+-------------+
| Step | Action                    | Time (in s) |
+------+---------------------------+-------------+
| 1    | default_platform          | 0           |
| 2    | is_ci                     | 0           |
| 3    | check_for_toolkit_updates | 5           |
| 4    | upload_to_play_store      | 18          |
+------+---------------------------+-------------+

[12:02:31]: fastlane.tools finished successfully 🎉
```

</details>

-----

@hichamboushaba me running the `bundle exec fastlane upload_metadata_to_play_store validate_only:true`, and as I explained to @AliSoftware above, means that we now have the update store listing on Play Store, which are currently in review...  🤷

Could you please check it out and let me know if everything is okay prior us hitting the `Publish` button on review approval. 🙏

AliSoftware commented:
> @AliSoftware that me adding the validate_only:true and running the lane actually updated the store listing, thus this parameter might not be doing what we thought it would... 🤷

That's because you provided the parameter to the command line invocation, instead of adding it as parameter to the `upload_to_play_store`call site of the action in your dummy lane in the Fastfile [like I suggested in my comment](https://github.com/woocommerce/woocommerce-android/pull/8439#issuecomment-1451840795).

`bundle exec fastlane somelane key1:value1 key2:value2` will call lane `somelane`, passing `{key1: value1, key2:  value2}` as the values for the `|options|` parameter to the `somelane` lane (i.e. as parameter to the `lane :somelane do |options|` block). But if you don't read that value (`options[:key1]` and `options[:key2]`) in that lane, that option will just go unused and do nothing. This is what happened here.
In particular, you didn't make use of that `options[:validate_only]` to pass it to the `validate_only:` parameter of the call site of the `upload_to_play_store` action that is called as part of your lane. So I don't know how you expected that value to be magically passed to the `upload_to_play_store` there.

Instead what I suggested you to do with my earlier comments **wasn't** to pass `validate_only:true` to the command line, **but** was to add `validate_only: true` to the `upload_to_play_store` call site. Like, edit your `Fastfile`, go to where you added your dummy lane (`lane :upload_metadata_to_play_store do …`), and in the implementation of that lane, edit the call to the `upload_to_play_store` action to add `validate_only:true` as an extra argument alongside all the other ones.

ParaskP7 commented:
> That's because you provided the parameter to the command line invocation, instead of adding it as parameter to the upload_to_play_storecall site of the action in your dummy lane in the Fastfile https://github.com/woocommerce/woocommerce-android/pull/8439#issuecomment-1451840795.

Me saying that [here](https://github.com/woocommerce/woocommerce-android/pull/8439#issuecomment-1451868412), and not getting feedback that what I was doing was wrong, incorrectly made me think that this would somehow work with fastlane/ruby magic... 😞 😞 😞

> bundle exec fastlane somelane key1:value1 key2:value2 will call lane somelane, passing {key1: value1, key2:  value2} as the values for the |options| parameter to the somelane lane (i.e. as parameter to the lane :somelane do |options| block). But if you don't read that value (options[:key1] and options[:key2]) in that lane, that option will just go unused and do nothing. This is what happened here.

👍 

> In particular, you didn't make use of that options[:validate_only] to pass it to the validate_only: parameter of the call site of the upload_to_play_store action that is called as part of your lane. So I don't know how you expected that value to be magically passed to the upload_to_play_store there.

Exactly, I for some reason magically expected that to happen, my bad... 😢 

> Instead what suggested you to do with my earlier comments was not to pass validate_only:true to the command line, but to add validate_only: true to the upload_to_play_store call site. Like, edit your Fastfile, go to where you added your dummy lane (lane :upload_metadata_to_play_store do …), and in the implementation of that lane, edit the call to the upload_to_play_store action to add validate_only:true as an extra argument alongside all the other ones.

I totally missed that and wasn't careful enough while wrangling other things and keep context switching to progress this PR, apologies... 😭

ParaskP7 commented:
**UPDATE**: I forgot to mention that here but I did incorporate the changes in this PR on the `12.6` finalization PR last Friday (see [here](https://github.com/woocommerce/woocommerce-android/pull/8530)). I then tried to run the release build but it failed (`C5ALGJYEP/p1678456705434639`). I then [reverted](https://github.com/woocommerce/woocommerce-android/pull/8530/commits/be083018945879e4bcec3b4318f1828feb9187f4) the change that enables the upload of both, metadata and changelogs, just to be able to complete the `12.6` finalization process.

FYI: I will this try again on the `12.7` finalization PR this Friday and give you another update on that. As Olivier mentioned, the failure might as well be due to some flakiness or race condition with the Google API, rather than anything else, so I'll just hit retry next time (in case that's what causing it). Cc @AliSoftware @oguzkocer @hichamboushaba

ParaskP7 commented:
**UPDATE**: FYI: Although this time the release build succeeded during `12.7` finalization, the automated metadata, or at least the automated releases notes upload didn't work as expected and this is (most probably) due to the retained `356` version code (for more info, see [PR comment](https://github.com/woocommerce/woocommerce-android/pull/8579#issuecomment-1473763694)). Cc @AliSoftware @oguzkocer @hichamboushaba

PS: Olivier, you might want to close this PR now as all of your changes have been applied via #8530 and #8579.
